### PR TITLE
 Create rust structs and replace bindgen structs

### PR DESCRIPTION
**Pull request by guibescos**
This PR :
- Creates rust structs in `c_oracle_header.rs` and replaces all the bindgen structs by the new structs.

### CODE REVIEW

File: program/rust/src/c_oracle_header.rs [hunk 1]
```diff
@@ -34,94 +34,137 @@ pub trait PythAccount: Pod {
...
+
+// Unsafe impl because product_list is of size 640 and there's no derived trait for this size
+unsafe impl Pod for MappingAccount {
+}
+unsafe impl Zeroable for MappingAccount {
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union CPubkey {
```
Comment (guibescos) [inline program/rust/src/c_oracle_header.rs:159]: Both fields are currently used in the code. I feel like we can get to a safe derivation if we :
- Replace CPubkey by the solana sdk Pubkey 
- Update functions to only use `k1_` or `k8_`.
Let's do it in a future PR.

File: program/rust/src/c_oracle_header.rs [hunk 2]
```diff
@@ -34,94 +34,137 @@ pub trait PythAccount: Pod {
...
-
-impl pc_pub_key_t {
-    pub fn new_unique() -> pc_pub_key_t {
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct PriceAccount {
+    pub magic_:          u32,
+    pub ver_:            u32,
+    pub type_:           u32,
+    pub size_:           u32,
```
Comment (Reisen) [inline program/rust/src/c_oracle_header.rs:61]: Can we move these in-line comments above the fields and make them Rust doc comments:

```suggestion
    /// Example documentation comment with a triple ///
    pub size_:           u32,
```

File: program/rust/src/c_oracle_header.rs [hunk 3]
```diff
@@ -34,94 +34,137 @@ pub trait PythAccount: Pod {
...
+    pub unused_3_:       i32,
+    pub prod_:           CPubkey, // Corresponding mapping account
+    pub next_:           CPubkey, // Next price account in the list
+    pub prev_slot_:      u64,     /* Second to last slot where aggregation was succesful
+                                   * (status : TRADING) */
+    pub prev_price_:     i64,       // Aggregate price at prev_slot_
+    pub prev_conf_:      u64,       // Confidence interval at prev_slot_
+    pub prev_timestamp_: i64,       // Timestamp of prev_slot_
+    pub agg_:            PriceInfo, // Last attempted aggregate results
+    pub comp_:           [PriceComponent; 32usize], // Publisher's price components
```
Comment (Reisen) [inline program/rust/src/c_oracle_header.rs:83]: Nitpick: Doesn’t need fixing but array sizes are always `usize` so can be specified as just 32.

File: program/rust/src/c_oracle_header.rs [hunk 4]
```diff
@@ -34,94 +34,137 @@ pub trait PythAccount: Pod {
...
+#[derive(Copy, Clone)]
+pub struct MappingAccount {
+    pub magic_:  u32,
+    pub ver_:    u32,
+    pub type_:   u32,
+    pub size_:   u32,
+    pub num_:    u32,
+    pub unused_: u32,
+    pub next_:   CPubkey,
+    pub prod_:   [CPubkey; 640usize],
```
Comment (Reisen) [inline program/rust/src/c_oracle_header.rs:141]: Can we lift the 640usize and others to a `const`, similar to the C headers. It’s easier to spot file-level constants as they show up as language items in IDEs.

File: program/rust/src/c_oracle_header.rs [hunk 5]
```diff
@@ -34,94 +34,137 @@ pub trait PythAccount: Pod {
     }
 }
 
-impl PythAccount for pc_map_table_t {
+impl PythAccount for MappingAccount {
     const ACCOUNT_TYPE: u32 = PC_ACCTYPE_MAPPING;
     const INITIAL_SIZE: u32 = PC_MAP_TABLE_T_PROD_OFFSET as u32;
```
Comment (Reisen) [inline program/rust/src/c_oracle_header.rs:39]: Can we not use `size_of::<MappingAccount>` here as in `ProductAccount`?

File: program/rust/src/c_oracle_header.rs (hunk 1 above)
Comment (Reisen) [reply program/rust/src/c_oracle_header.rs:159]: Agree with the first suggestion, we should use the solana `Pubkey` type directly in a future PR. The second suggestion would just keep us using the ad-hoc unsafe functions we should scrap.

File: program/rust/src/c_oracle_header.rs (hunk 3 above)
Comment (guibescos) [reply program/rust/src/c_oracle_header.rs:83]: In the future they'll be constants

File: program/rust/src/c_oracle_header.rs (hunk 5 above)
Comment (guibescos) [reply program/rust/src/c_oracle_header.rs:39]: No, added a comment explaning.

File: program/rust/src/c_oracle_header.rs (hunk 4 above)
Comment (guibescos) [reply program/rust/src/c_oracle_header.rs:141]: Yes, I agree. Next PR will be getting the constants from oracle.h . As of now, I'll use the C constant for this PR.

File: program/rust/src/c_oracle_header.rs (hunk 2 above)
Comment (guibescos) [reply program/rust/src/c_oracle_header.rs:61]: Good

**Review by Reisen — APPROVED** [review]

**Pull request merged**